kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,608 @@
+
+## Motivation
+
+Kubernetes provides a declarative resource model that core components (scheduler and kubelet) honor to behave consistently and satisfy QoS guarantees.
+However, using this model can lead to low-utilization of the cluster for the following reasons:
+
+1. It is hard for users to estimate accurate resource usage for applications. Also, users may not understand the resource model and not set it at all.
+2. The default in-tree scheduling plugins (Score) that Kubernetes provides don't consider live node utilization values.
+
+This proposal utilizes real time resource usage to schedule pods with proposed plugins. The eventual goal is to increase cluster utilization and decrease costs of cluster management without breaking the contract of Kubernetes Resource Model.
+
+### Goals
+
+1. Provide configurable scheduling plugins to increase cluster utilization.
+2. Expected CPU Utilisation per node should not go beyond X%.
+3. Instances of an app should spread across failure domains. The failure domain would be host at the lowest level and can extend to racks, zones, regions, etc.
+4. To not affect the behavior of default score plugins unless necessary.
+5. Implement the above features as Score plugins
+
+### Non-Goals
+
+1. Implement the above constraints (2, 3) as Filter plugins.
+2. Descheduling due to unexpected outcome (hot nodes, fragmentation etc.) of past scoring by plugins is not addressed in the initial design.
+3. Memory, Network, and Disk utilization are not considered in the initial design.
+
+## Proposal
+
+
+### User Stories
+
+
+#### Story 1
+
+
+As a company relying on the public cloud, we would like to minimize machine costs by efficiently utilizing the nodes leased.
+
+
+As a company-owned data center, we would like to minimize the hardware and maintenance associated with clusters by getting the maximum possible juice out of existing machines.
+
+
+#### Story 2
+
+Increasing resource utilization as much as possible may not be the right solution for all clusters. As it always takes some time to scale up a cluster to handle the sudden spikes of load, cluster admins would like to leave adequate room for the bursty load, so there is enough time to add more nodes to the cluster.
+
+
+### Notes/Constraints/Caveats
+
+Enabling our plugin(s) will cause conflict with 2 default scoring plugins: "NodeResourcesLeastAllocated" and "NodeResourcesBalancedAllocation" plugins. So it is strongly advised to disable them when enabling plugin(s) mentioned in this proposal.
+For the 3rd Goal, since we utilise PodTopologySpread plugin scoring and to prevent double scoring, it is recommended to disable "PodTopologySpread" Score plugin only. The motivation behind using this plugin is explained under "Spreading with BestFitBinPack".
+
+
+### Risks and Mitigations
+
+If utilization metrics are not available for a long time, we will fall back to the best fit bin pack based on allocations. There is no user action needed for this.
```

Comment:
nit: Add reference to where this is analyzed. `#bad-metrics`?

Reply:
in this case BestFitBinPack based just on allocations seems similar to the existing `NodeResourcesMostAllocated` (where NRMA is just BFBP with a target utilization of 100%). If that's the case, maybe we could look at just extending NRMA to have a configurable target utilization + live metrics, rather than introducing a new plugin?

Though I may be oversimplifying it, but just a thought

Reply:
Since this is more like a fall-back in case of failures, and it would be utilized by multiple types of "load aware" plugins (Both SafeBalancing and BestFitBinPack in this PR) extending `NodeResourcesMostAllocated` may not be a good idea. Perhaps falling back to `default-scheduler` profile is one way to achieve this, if real load plugins can not work.

Reply:
Could you still link to `#bad-metrics` so readers can know that there is more to this topic?

Reply:
Done

Reply:
> The architecture required to obtain and cache the metrics might be a non-neglibile risk, even if the feature is disabled. I prefer this evolves independently first. But we could upstream the "target utilization" bits and somehow reuse that code in here.

For some reason this newer comment is shown after my older comment and I missed this. Yeah we don't have a clear implementation yet or design for how the fall back works, it adds quite a bit of complexity -> "for example when does fall back kick in, is it for all cluster nodes or just a few nodes?"
We would like to analyse this further with time and see the best solution. If you want I can remove this part from KEP.

Reply:
@Huang-Wei your thoughts on this?

Reply:
I agree that you can build this standalonely. However the ideal case is to decouple the implementations can reuse the commons as much as possible, specifically:

1. Abstract the metric source - like in the upstream, it's from the pod's spec and cluster's initial capacity; while in this plugin, it's from realtime metrics
1. With the metric source serving as an contract/interface, we can build a more common "utilization" based algorithm, which can be used by both upstream and the plugins discussed here.

Maybe you can mention this in the KEP, but up to you:)